frame/contracts/src/account_db.rs

```diff
@@ -210,18 +218,70 @@ impl<T: Trait> AccountDb<T> for DirectAccountDb {
 					new_info.last_write = Some(<frame_system::Module<T>>::block_number());
 				}
 
-				for (k, v) in changed.storage.into_iter() {
-					if let Some(value) = child::get_raw(
-						&new_info.child_trie_info(),
-						&blake2_256(&k),
-					) {
-						new_info.storage_size -= value.len() as u32;
+				// NB: this call allocates internally. To keep allocations to the minimum we cache
+				// the child trie info here.
+				let child_trie_info = new_info.child_trie_info();
+
+				// Here we iterate over all storage key-value pairs that were changed throughout the
+				// execution of a contract and apply them to the substrate storage.
+				for (key, opt_new_value) in changed.storage.into_iter() {
+					let hashed_key = blake2_256(&key);
+
+					// In order to correctly update the book keeping we need to fetch the previous
+					// value of the key-value pair.
+					//
+					// It might be a bit more clean if we had an API that supported getting the size
+					// of the value without going through the loading of it. But at the moment of
+					// writing, there is no such API.
+					//
+					// That's not a show stopper in any case, since the performance cost is
+					// dominated by the trie traversal anyway.
+					let opt_prev_value = child::get_raw(&child_trie_info, &hashed_key);
+
+					// Update the total number of KV pairs and the number of empty pairs.
+					match (&opt_prev_value, &opt_new_value) {
+						(Some(prev_value), None) => {
+							new_info.total_pair_count -= 1;
+							if prev_value.is_empty() {
+								new_info.empty_pair_count -= 1;
+							}
+						},
+						(None, Some(new_value)) => {
+							new_info.total_pair_count += 1;
+							if new_value.is_empty() {
+								new_info.empty_pair_count += 1;
+							}
+						},
+						(Some(prev_value), Some(new_value)) => {
+							if prev_value.is_empty() {
+								new_info.empty_pair_count -= 1;
+							}
+							if new_value.is_empty() {
+								new_info.empty_pair_count += 1;
+							}
+						}
+						(None, None) => {}
 					}
-					if let Some(value) = v {
-						new_info.storage_size += value.len() as u32;
-						child::put_raw(&new_info.child_trie_info(), &blake2_256(&k), &value[..]);
-					} else {
-						child::kill(&new_info.child_trie_info(), &blake2_256(&k));
+
+					// Update the total storage size.
+					let prev_value_len = opt_prev_value
+						.as_ref()
+						.map(|old_value| old_value.len() as u32)
+						.unwrap_or(0);
+					let new_value_len = opt_new_value
+						.as_ref()
+						.map(|new_value| new_value.len() as u32)
+						.unwrap_or(0);
+					new_info.storage_size = new_info
+						.storage_size
+						.saturating_sub(prev_value_len)
+						.saturating_add(new_value_len);
```

Comment:
We are unsigned. You need to swap those in order to prevent the value to saturate to zero in between.

Reply:
Sure! But fwiw note that we must never underflow here, it must be guaranteed by the foregoing logic.

Reply:
Did not know that. Still think that is makes sense to swap it for clarity.